File: langs/csharp/src/VMx/Commands/ModeledCrudCommands.cs
using System.Reactive;
using System.Reactive.Disposables;
using System.Windows.Input;

namespace VMx.Commands;

/// <summary>
/// Composes the three modeled-CRUD commands (Create/Update/Delete-current) against
/// a current-VM provider and per-action callbacks. See spec/06-composite-vm.md
/// §Modeled CRUD commands and ADR-0016.
/// </summary>
public sealed class ModeledCrudCommands<M, VM> : IDisposable
    where VM : class
{
    /// <summary>Command that invokes the create-new action.</summary>
    public ICommand CreateNewCommand { get; }

    /// <summary>Command that invokes update_current with the current VM. CanExecute requires current != null.</summary>
    public ICommand UpdateCurrentCommand { get; }

    /// <summary>Command that invokes delete_current with the current VM. CanExecute requires current != null.</summary>
    public ICommand DeleteCurrentCommand { get; }

    private readonly CompositeDisposable _disposables = new();

    /// <summary>Creates a new CRUD command set.</summary>
    /// <param name="current">Provider returning the current VM (or null).</param>
    /// <param name="createNew">Action invoked by CreateNewCommand.</param>
    /// <param name="updateCurrent">Action invoked by UpdateCurrentCommand with the current VM.</param>
    /// <param name="deleteCurrent">Action invoked by DeleteCurrentCommand with the current VM.</param>
    /// <param name="confirmUpdate">Optional async confirm gate for update.</param>
    /// <param name="confirmDelete">Optional async confirm gate for delete.</param>
    /// <param name="currentChanged">
    /// Optional trigger that fires whenever the value returned by
    /// <paramref name="current"/> changes (typically the owning CompositeVM's
    /// current-child-changed stream). When supplied, it is wired as a trigger on
    /// the Update and Delete commands so a bound button's enabled state refreshes
    /// the moment the selection changes — otherwise the predicate
    /// (<c>current() is not null</c>) has no way to raise
    /// <see cref="ICommand.CanExecuteChanged"/> and the button goes stale (VMX-011).
    /// </param>
    public ModeledCrudCommands(
        Func<VM?> current,
        Action createNew,
        Action<VM> updateCurrent,
        Action<VM> deleteCurrent,
        Func<Task<bool>>? confirmUpdate = null,
        Func<Task<bool>>? confirmDelete = null,
        IObservable<Unit>? currentChanged = null)
    {
        var create = RelayCommand.Builder().Task(createNew).Build();
        var updateBuilder = RelayCommand.Builder()
            .Task(() => { var c = current(); if (c is not null) updateCurrent(c); })
            .Predicate(() => current() is not null);
        var deleteBuilder = RelayCommand.Builder()
            .Task(() => { var c = current(); if (c is not null) deleteCurrent(c); })
            .Predicate(() => current() is not null);
        if (currentChanged is not null)
        {
            updateBuilder = updateBuilder.Triggers(currentChanged);
            deleteBuilder = deleteBuilder.Triggers(currentChanged);
        }

        var update = updateBuilder.Build();
        var delete = deleteBuilder.Build();
        _disposables.Add((IDisposable)create);
        _disposables.Add((IDisposable)update);
        _disposables.Add((IDisposable)delete);

        CreateNewCommand = create;
        if (confirmUpdate is not null)
        {
            var wrapped = new ConfirmationDecoratorCommand(update, confirmUpdate);
            _disposables.Add(wrapped);
            UpdateCurrentCommand = wrapped;
        }
        else
        {
            UpdateCurrentCommand = update;
        }
        if (confirmDelete is not null)
        {
            var wrapped = new ConfirmationDecoratorCommand(delete, confirmDelete);
            _disposables.Add(wrapped);
            DeleteCurrentCommand = wrapped;
        }
        else
        {
            DeleteCurrentCommand = delete;
        }
    }

    /// <summary>Disposes the underlying commands. Idempotent.</summary>
    public void Dispose() => _disposables.Dispose();
}

File: langs/python/src/vmx/commands/modeled_crud_commands.py
"""ModeledCrudCommands — Create / UpdateCurrent / DeleteCurrent helper.

See spec/06-composite-vm.md §Modeled CRUD commands and ADR-0016.
"""

from __future__ import annotations

from collections.abc import Awaitable, Callable
from typing import Generic, TypeVar

from vmx.commands.confirmation_decorator_command import ConfirmationDecoratorCommand
from vmx.commands.protocols import Command
from vmx.commands.relay_command import RelayCommand

VM = TypeVar("VM")
M = TypeVar("M")


class ModeledCrudCommands(Generic[M, VM]):
    """Three CRUD commands wired against a current-VM provider."""

    create_new_command: Command
    update_current_command: Command
    delete_current_command: Command

    def __init__(
        self,
        current: Callable[[], VM | None],
        create_new: Callable[[], None],
        update_current: Callable[[VM], None],
        delete_current: Callable[[VM], None],
        confirm_update: Callable[[], Awaitable[bool]] | None = None,
        confirm_delete: Callable[[], Awaitable[bool]] | None = None,
    ) -> None:
        def _do_update() -> None:
            c = current()
            if c is not None:
                update_current(c)

        def _do_delete() -> None:
            c = current()
            if c is not None:
                delete_current(c)

        create = RelayCommand.builder().task(create_new).build()
        update = (
            RelayCommand.builder().task(_do_update).predicate(lambda: current() is not None).build()
        )
        delete = (
            RelayCommand.builder().task(_do_delete).predicate(lambda: current() is not None).build()
        )

        # Track the inner RelayCommands so dispose() can complete their
        # can_execute_changed subjects (parity with C# ModeledCrudCommands.
        # Dispose). No triggers are wired here, so there are no trigger
        # subscriptions to release.
        self._inner_relays = (create, update, delete)
        self._disposed = False

        self.create_new_command = create
        self.update_current_command = (
            ConfirmationDecoratorCommand(update, confirm=confirm_update)
            if confirm_update is not None
            else update
        )
        self.delete_current_command = (
            ConfirmationDecoratorCommand(delete, confirm=confirm_delete)
            if confirm_delete is not None
            else delete
        )

    def dispose(self) -> None:
        """Dispose the inner RelayCommands and any confirmation wrappers.

        Idempotent: subsequent calls are a no-op.

        The public update/delete commands may be ``ConfirmationDecoratorCommand``
        wrappers (when ``confirm_update`` / ``confirm_delete`` are supplied). Each
        wrapper owns an ``errors`` ``Subject`` whose contract is to complete on
        ``dispose``, so the wrappers are disposed here alongside the inner relays
        — parity with C# ``ModeledCrudCommands.Dispose``, which adds the wrappers
        to its disposables. Each distinct command is disposed exactly once (when
        no confirm hook is supplied the public command *is* the inner relay).
        """
        if self._disposed:
            return
        self._disposed = True
        # The inner relays always own disposable state. create_new_command is
        # always the inner create relay; update/delete may additionally be a
        # ConfirmationDecoratorCommand wrapper whose errors Subject must complete.
        disposables: list[RelayCommand | ConfirmationDecoratorCommand] = list(self._inner_relays)
        for command in (self.update_current_command, self.delete_current_command):
            if isinstance(command, ConfirmationDecoratorCommand):
                disposables.append(command)
        first_error: BaseException | None = None
        for cmd in disposables:
            try:
                cmd.dispose()
            except BaseException as error:
                if first_error is None:
                    first_error = error
        if first_error is not None:
            raise first_error
